pre-commit/check.py

```diff
@@ -0,0 +1,161 @@
+    """return all files that are committed but not pushed."""
+    # get upstream branch name
+    upstream_branch = (
+        os.popen("git rev-parse --abbrev-ref --symbolic-full-name @{u}").read().strip()
+    )
+    # get all commits that are on the current branch but not on the upstream branch
+    commits = os.popen(f"git log {upstream_branch}..HEAD --oneline").read().strip()
+    commits = [x.split()[0] for x in commits.split("\n")]
+    # get all files that are changed in the commits
+    files: List[Path] = []
+    for commit in commits:
+        files.extend(get_git_commit_files(commit))
+    # remove duplicates
+    return sorted(list(set(files)))
+
+
+# check if PRECOMMIT_ENABLE_SLOW is set to true
+if os.environ.get("PRE_COMMIT_ENABLE_SLOW", "false") != "true":
+    print("Skipping slow checks.")
+    print("Set PRE_COMMIT_ENABLE_SLOW=true to activate.")
+    exit(0)
+
+PROJECT_NAME = "determined"
+root = Path(os.getenv("PROJECT_ROOT", os.getcwd())).absolute()
+if not str(root).endswith(PROJECT_NAME) and not str(root).endswith("saas"):  # FIXME
+    print(
+        "Please run this script from the root of the project or set it using "
+        + "PROJECT_ROOT env variable"
+    )
+    exit(1)
+os.chdir(root)
+
+rules: Dict[Path, Union[str, List[str]]] = {
+    root
+    / "harness": [
+        "make -j fmt; make -j check",
+        "make -j build",
+    ],
+    root
+    / "proto": [
+        "make fmt check build",
+        "make -C ../bindings build && make -C ../webui/react bindings-copy-over && make -C ../webui/react check",
+    ],
+    root / "webui" / "react": ["make -j fmt; make -j check", "make -j test && make -j build"],
+    root
+    / "master": [
+        "make -C ../proto build",
+        "make build",
+        "make fmt; make check",
+        "make test",
+    ],
+    root / "docs": "make fmt check build",
+    root / ".circleci": "circleci config validate config.yml",
+    root / "e2e_tests": "make fmt check",
+    root / "model_hub": "make fmt check",
+}
+
+
+# check if path is the same or child of one of the rules and execute the rule as
+# subprocess
+def run_rule(rule_path: Path) -> int:
+    rule = rules[rule_path]
+    os.chdir(rule_path)
+    print(f'in direcotry "{rule_path.relative_to(root)}" run: {rule}')
+    cmds = rule if isinstance(rule, list) else [rule]
+    for cmd in cmds:
+        assert isinstance(cmd, str)
+        return_code = os.system(cmd)
+        if return_code != 0:
+            print(f'command "{cmd}" failed with return code {return_code}', file=sys.stderr)
+            return return_code
+    return 0
+
+
+def find_rules(paths: List[Path]):
+    resolved_paths = set()
+    for dirty_path in paths:
+        for rule_path in rules.keys():
+            if is_child(dirty_path, rule_path):
+                resolved_paths.add(rule_path)
+    return resolved_paths
+
+
+def main():
+    # define an arg to get the stage
+    argparser = argparse.ArgumentParser()
```

Comment:
For this application, I think we can strip down argument parsing quite a bit. If we just expect a list of file args as arguments, I think we can condense the start of this function to something like:

```python
changed_files = [Path(arg).absolute() for arg in sys.argv]
if len(changed_files == 0):
  print("no changed files")
  exit(0)
```

Reply:
I agree but IMO the existing one is a bit more extensible, gives us nice err msgs, help etc and makes it easier for me to have my old version of this where it's not used through `pre-commit`. keep?